src/Microsoft.ComponentDetection.Common/DirectoryWalkerFactory.cs

```diff
@@ -0,0 +1,128 @@
+﻿namespace Microsoft.ComponentDetection.Common;
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Enumeration;
+using System.Linq;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using Microsoft.ComponentDetection.Contracts;
+using Microsoft.ComponentDetection.Contracts.Internal;
+using Microsoft.Extensions.Logging;
+
+public class DirectoryWalkerFactory : IDirectoryWalkerFactory
+{
+    private readonly IPathUtilityService pathUtilityService;
+    private readonly ILogger<DirectoryWalkerFactory> logger;
+
+    public DirectoryWalkerFactory(IPathUtilityService pathUtilityService, ILogger<DirectoryWalkerFactory> logger)
+    {
+        this.logger = logger;
+        this.pathUtilityService = pathUtilityService;
+    }
+
+    public async Task WalkDirectoryAsync(DirectoryInfo root, ExcludeDirectoryPredicate directoryExclusionPredicate, IComponentRecorder recorder, Func<ProcessRequest, Task> callback, IEnumerable<string> filePatterns)
+    {
+        if (!root.Exists)
+        {
+            this.logger.LogError("Root directory doesn't exist: {RootFullName}", root.FullName);
+            return;
+        }
+
+        PatternMatchingUtility.FilePatternMatcher fileIsMatch = null;
+
+        if (filePatterns == null || !filePatterns.Any())
+        {
+            fileIsMatch = _ => true;
+        }
+        else
+        {
+            fileIsMatch = PatternMatchingUtility.GetFilePatternMatcher(filePatterns);
+        }
+
+        var actionBlock = new ActionBlock<ProcessRequest>(callback, new ExecutionDataflowBlockOptions() { MaxDegreeOfParallelism = Environment.ProcessorCount });
+
+        var walker = new FileSystemEnumerable<FileSystemInfo>(
+            root.FullName,
+            (ref FileSystemEntry entry) => entry.ToFileSystemInfo(),
+            new EnumerationOptions
+            {
+                AttributesToSkip = FileAttributes.Hidden | FileAttributes.System,
```

Comment:
I think we might actually need to scan hidden directories.

Reply:
We currently do not. The original walker does not set these options, and `FileAttributes.Hidden | FileAttributes.System` is [the default](https://learn.microsoft.com/en-us/dotnet/api/system.io.enumerationoptions.attributestoskip?view=net-7.0#property-value).

https://github.com/microsoft/component-detection/blob/1c3ef853d24315f8648ec3b0b38f776b927af0f4/src/Microsoft.ComponentDetection.Common/FastDirectoryWalkerFactory.cs#L93-L98